### DISCUSSION TITLE
liquidsoap https issue after passing Icecast2 in SSL

### DISCUSSION THREAD

**Discussion by nicolasdeslilas** [savonet/liquidsoap#3364] [User Support]
Hello,
**Describe the bug**
Since I passed Icecast2 in HTTPS I can't send from Liquidsoap to icecast2 I have:
CCertificate[2] subject=/C=US/O=Internet Security Research Group/CN=ISRG Root X1
Certificate[2] issuer =/O=Digital Signature Trust Co./CN=DST Root CA X3
SSL: rejecting connection - error=20

**Expected behavior**
I just wanna send a liquidsoap stream on icecast adresse like https://3615.world:8000

- OS: [e.g. Debian, OSX]
Ubuntu 22

Thanks for your help !

**Comment by vitoyucepi** [comment]
Hi @nicolasdeslilas,
Could you share more information about your installation?
1. What's the version of liquidsoap you're using?
1. How did you install the liquidsoap, packages, container, opam, or something else?
1. How have you configured the `output.harbor`?
   Could you paste a code snippet?

**Comment by nicolasdeslilas** [comment]
Hi @vitoyucepi,

I use Liquidsoap 1.4.1 but I also try with a 2.2.0 with a docker.
I thin I add installed liquidsoap by apt install

Here a sample of my code:

`
live = audio_to_stereo(input.https("https://localhost:8000/stream"))



stream = fallback(track_sensitive=false,[live, radio])

stream = fallback(track_sensitive=false, [strip_blank(max_blank=5., stream), radio])


output.icecast(%mp3(bitrate=192),
    host = "78.194.135.9", port = 8000,
    name = "3615 Radio", description = "Happy Mix", genre = "All Style you like", url = "https://3615.world",
    password = "xxxx",
    insecure=true,
    mount = "listen", stream)`

and I tying somthing like this too:

`set("harbor.ssl.certificate","/etc/letsencrypt/live/3615.world/fullchain.pem")
set("harbor.ssl.private_key","/etc/letsencrypt/live/3615.world/privkey.pem")

live = audio_to_stereo(input.harbor("localhost",port=8000,password="xxxx"))`

Thanks for your help !

**Comment by vitoyucepi** [comment]
Let's stick to 2.2.0.
```ruby
radio = sine()
live = input.harbor(
  "live",
  port=8000,
  transport=http.transport.ssl(
    certificate="/etc/letsencrypt/live/3615.world/fullchain.pem",
    key="/etc/letsencrypt/live/3615.world/privkey.pem",
  ),
)
stream = fallback([live, radio], track_sensitive=false)
output.icecast(
  %mp3(bitrate=192),
  stream,
  host="78.194.135.9",
  port=8000,
  transport=http.transport.ssl(),
  user="source",
  password="hackme",
  mount="listen",
  name="3615 Radio",
  description="Happy Mix",
  genre="All Style you like",
  url="https://3615.world",
)
```
1. `input.harbor` will listen for encrypted connections on port 8000.
1. `output.icecast` will stream over an encrypted connection to a remote server.

**Comment by toots** [comment]
Thanks for the support @vitoyucepi.

Yes, there are two different operations, sending and receiving. For receiving, the server has a certificate with a private key and tells the client that it can serve a given domain with the associated certificate. Then the client decides if the certificate is trust worthy based on its list of trusted ones.

The problem seems to be that the host that is sending to icecast does not trust the SSL certificate returned by the icecast server. If nothing works, you always have the option to add it to the SSL recognized certificates:
 
See: https://www.liquidsoap.info/doc-2.2.0/reference.html#http.transport.ssl
> `certificate` (of type `string?`, which defaults to `null`): Path to certificate file. Required in server mode, e.g. `input.harbor`, etc. If passed in client mode, certificate is added to the list of valid certificates.

But before that I would suggest to see why your host does not trust this certificate. Also,  you can try to use the TLS implementation which might yield different results.

**Comment by nicolasdeslilas** [comment]
Hello Team !

All is working fine now with this conf:

live = input.harbor(
  "live",
  port=8000,
  transport=http.transport.ssl(
    certificate="/etc/letsencrypt/live/3615.world/fullchain.pem",
    key="/etc/letsencrypt/live/3615.world/privkey.pem",
  ),
)
stream = fallback([live, radio], track_sensitive=false)
output.icecast(
  %mp3(bitrate=192),
  stream,
  host="78.194.135.9",
  port=8000,
  transport=http.transport.ssl(),
  user="source",
  password="hackme",
  mount="listen",
  name="3615 Radio",
  description="Happy Mix",
  genre="All Style you like",
  url="https://3615.world",
)

Thanks for your reactivity and your Help !
Love
